File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis.VisualBasic.Symbols/CrefTypeParameterSymbol.cs
using System.Collections.Immutable;
using Microsoft.CodeAnalysis.VisualBasic.Syntax;
using Microsoft.VisualBasic.CompilerServices;
using Roslyn.Utilities;

namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
{
	internal sealed class CrefTypeParameterSymbol : TypeParameterSymbol
	{
		private readonly int _ordinal;

		private readonly string _name;

		private readonly SyntaxReference _syntaxReference;

		public override int Ordinal => _ordinal;

		public override string Name => _name;

		public override bool HasConstructorConstraint => false;

		public override bool HasReferenceTypeConstraint => false;

		public override bool HasValueTypeConstraint => false;

		internal override ImmutableArray<TypeSymbol> ConstraintTypesNoUseSiteDiagnostics => ImmutableArray<TypeSymbol>.Empty;

		public override Symbol ContainingSymbol => null;

		public override ImmutableArray<SyntaxReference> DeclaringSyntaxReferences => ImmutableArray.Create(_syntaxReference);

		public override ImmutableArray<Location> Locations => ImmutableArray.Create(_syntaxReference.GetLocation());

		public override TypeParameterKind TypeParameterKind => TypeParameterKind.Cref;

		public override VarianceKind Variance => VarianceKind.None;

		public CrefTypeParameterSymbol(int ordinal, string name, TypeSyntax syntax)
		{
			_ordinal = ordinal;
			_name = name;
			_syntaxReference = syntax.GetReference();
		}

		public override ImmutableArray<VisualBasicAttributeData> GetAttributes()
		{
			return ImmutableArray<VisualBasicAttributeData>.Empty;
		}

		internal override ImmutableArray<TypeParameterConstraint> GetConstraints()
		{
			return ImmutableArray<TypeParameterConstraint>.Empty;
		}

		internal override void ResolveConstraints(ConsList<TypeParameterSymbol> inProgress)
		{
		}

		internal override void EnsureAllConstraintsAreResolved()
		{
		}

		public override bool Equals(TypeSymbol other, TypeCompareKind comparison)
		{
			return Equals(other as CrefTypeParameterSymbol);
		}

		public bool Equals(CrefTypeParameterSymbol other)
		{
			if ((object)this == other)
			{
				return true;
			}
			if ((object)other == null)
			{
				return false;
			}
			return EmbeddedOperators.CompareString(_name, other._name, TextCompare: false) == 0 && _ordinal == other._ordinal && _syntaxReference.GetSyntax().Equals(other._syntaxReference.GetSyntax());
		}

		public override int GetHashCode()
		{
			return Hash.Combine(_name, _ordinal);
		}
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/VisualBasic/Symbols/CrefTypeParameterSymbol.vb
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
Imports Microsoft.VisualBasic.CompilerServices
Imports Roslyn.Utilities
Imports System
Imports System.Collections.Immutable
Imports System.Threading

Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
	Friend NotInheritable Class CrefTypeParameterSymbol
		Inherits TypeParameterSymbol
		Private ReadOnly _ordinal As Integer

		Private ReadOnly _name As String

		Private ReadOnly _syntaxReference As SyntaxReference

		Friend Overrides ReadOnly Property ConstraintTypesNoUseSiteDiagnostics As ImmutableArray(Of TypeSymbol)
			Get
				Return ImmutableArray(Of TypeSymbol).Empty
			End Get
		End Property

		Public Overrides ReadOnly Property ContainingSymbol As Symbol
			Get
				Return Nothing
			End Get
		End Property

		Public Overrides ReadOnly Property DeclaringSyntaxReferences As ImmutableArray(Of SyntaxReference)
			Get
				Return ImmutableArray.Create(Of SyntaxReference)(Me._syntaxReference)
			End Get
		End Property

		Public Overrides ReadOnly Property HasConstructorConstraint As Boolean
			Get
				Return False
			End Get
		End Property

		Public Overrides ReadOnly Property HasReferenceTypeConstraint As Boolean
			Get
				Return False
			End Get
		End Property

		Public Overrides ReadOnly Property HasValueTypeConstraint As Boolean
			Get
				Return False
			End Get
		End Property

		Public Overrides ReadOnly Property Locations As ImmutableArray(Of Location)
			Get
				Return ImmutableArray.Create(Of Location)(Me._syntaxReference.GetLocation())
			End Get
		End Property

		Public Overrides ReadOnly Property Name As String
			Get
				Return Me._name
			End Get
		End Property

		Public Overrides ReadOnly Property Ordinal As Integer
			Get
				Return Me._ordinal
			End Get
		End Property

		Public Overrides ReadOnly Property TypeParameterKind As Microsoft.CodeAnalysis.TypeParameterKind
			Get
				Return Microsoft.CodeAnalysis.TypeParameterKind.Cref
			End Get
		End Property

		Public Overrides ReadOnly Property Variance As VarianceKind
			Get
				Return VarianceKind.None
			End Get
		End Property

		Public Sub New(ByVal ordinal As Integer, ByVal name As String, ByVal syntax As TypeSyntax)
			MyBase.New()
			Me._ordinal = ordinal
			Me._name = name
			Me._syntaxReference = syntax.GetReference()
		End Sub

		Friend Overrides Sub EnsureAllConstraintsAreResolved()
		End Sub

		Public Overrides Function Equals(ByVal other As TypeSymbol, ByVal comparison As TypeCompareKind) As Boolean
			Return Me.Equals(TryCast(other, CrefTypeParameterSymbol))
		End Function

		Public Function Equals(ByVal other As CrefTypeParameterSymbol) As Boolean
			Dim flag As Boolean
			Dim flag1 As Boolean
			If (CObj(Me) = CObj(other)) Then
				flag = True
			ElseIf (other IsNot Nothing) Then
				If (EmbeddedOperators.CompareString(Me._name, other._name, False) <> 0 OrElse Me._ordinal <> other._ordinal) Then
					flag1 = False
				Else
					Dim syntaxReference As Microsoft.CodeAnalysis.SyntaxReference = Me._syntaxReference
					Dim cancellationToken As System.Threading.CancellationToken = New System.Threading.CancellationToken()
					Dim syntax As SyntaxNode = syntaxReference.GetSyntax(cancellationToken)
					Dim syntaxReference1 As Microsoft.CodeAnalysis.SyntaxReference = other._syntaxReference
					cancellationToken = New System.Threading.CancellationToken()
					flag1 = syntax.Equals(syntaxReference1.GetSyntax(cancellationToken))
				End If
				flag = flag1
			Else
				flag = False
			End If
			Return flag
		End Function

		Public Overrides Function GetAttributes() As ImmutableArray(Of VisualBasicAttributeData)
			Return ImmutableArray(Of VisualBasicAttributeData).Empty
		End Function

		Friend Overrides Function GetConstraints() As ImmutableArray(Of TypeParameterConstraint)
			Return ImmutableArray(Of TypeParameterConstraint).Empty
		End Function

		Public Overrides Function GetHashCode() As Integer
			Return Hash.Combine(Of String)(Me._name, Me._ordinal)
		End Function

		Friend Overrides Sub ResolveConstraints(ByVal inProgress As ConsList(Of TypeParameterSymbol))
		End Sub
	End Class
End Namespace